File: src/main/java/org/tdf/lotusvm/runtime/HostFunction.java
package org.tdf.lotusvm.runtime;

import org.tdf.lotusvm.ModuleInstance;
import org.tdf.lotusvm.types.FunctionType;

import java.util.*;

public abstract class HostFunction implements FunctionInstance {
    private final FunctionType type;
    private final String name;
    ModuleInstanceImpl instance;
    private Set<String> alias;

    public HostFunction(String name, FunctionType type) {
        this.name = name;
        this.type = type;
    }

    public HostFunction(String name, FunctionType type, String... alias) {
        this.name = name;
        this.type = type;
        this.alias = new HashSet<>();
        this.alias.addAll(Arrays.asList(alias));
    }

    public Set<String> getAlias() {
        if (alias == null)
            return Collections.emptySet();
        return alias;
    }

    protected ModuleInstance getInstance() {
        return instance;
    }

    @Override
    public int parametersLength() {
        return getType().getParameterTypes().size();
    }

    @Override
    public int getArity() {
        return getType().getResultTypes().size();
    }

    @Override
    public FunctionType getType() {
        return type;
    }


    @Override
    public abstract long execute(long[] parameters);

    @Override
    public boolean isHost() {
        return true;
    }

    public String getName() {
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        HostFunction that = (HostFunction) o;
        return Objects.equals(name, that.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    protected void putMemory(int offset, byte[] data) {
        instance.memory.put(offset, data);
    }

    protected byte[] loadMemory(int offset, int length) {
        return instance.memory.load(offset, length);
    }
}

File: src/main/java/org/tdf/lotusvm/runtime/HostFunction.kt
package org.tdf.lotusvm.runtime

import org.tdf.lotusvm.ModuleInstance
import org.tdf.lotusvm.types.FunctionType
import java.util.*

abstract class HostFunction(
    val name: String,
    override val type: FunctionType,
    vararg alias: String
) : FunctionInstance {

    lateinit var instance: ModuleInstance

    val alias: MutableSet<String> = mutableSetOf()

    init {
        this.alias.addAll(listOf(*alias))
    }

    protected val memory: Memory
        get() = instance.memory

    override val paramSize: Int
        get() = type.parameterTypes.size
    override val arity: Int
        get() = type.resultTypes.size

    abstract override fun execute(args: LongArray): Long
    override val isHost: Boolean
        get() = true

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as HostFunction
        return name == that.name
    }

    override fun hashCode(): Int {
        return Objects.hash(name)
    }
}